File: src/liveplugin/toolwindow/addplugin/AddNewPluginAction.java
package liveplugin.toolwindow.addplugin;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.InputValidatorEx;
import com.intellij.openapi.ui.Messages;
import liveplugin.IDEUtil;
import liveplugin.Icons;
import liveplugin.LivePluginAppComponent;
import liveplugin.pluginrunner.GroovyPluginRunner;
import liveplugin.toolwindow.RefreshPluginsPanelAction;
import liveplugin.toolwindow.util.PluginsIO;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;

@SuppressWarnings("ComponentNotRegistered")
public class AddNewPluginAction extends AnAction implements DumbAware {
	private static final Logger log = Logger.getInstance(AddNewPluginAction.class);
	private static final String addNewPluginTitle = "Add New Plugin";

	public AddNewPluginAction() {
		super("New Plugin", "Create new plugin", Icons.NEW_PLUGIN_ICON);
	}

	@Override public void actionPerformed(@NotNull AnActionEvent event) {
		String newPluginId = Messages.showInputDialog(
				event.getProject(),
				"Enter new plugin name:",
				addNewPluginTitle,
				null, "", new PluginIdValidator()
		);
		if (newPluginId == null) return;

		try {

			String text = LivePluginAppComponent.defaultPluginScript();
			PluginsIO.createFile(LivePluginAppComponent.pluginsRootPath() + "/" + newPluginId, GroovyPluginRunner.MAIN_SCRIPT, text);

		} catch (IOException e) {
			Project project = event.getProject();
			if (project != null) {
				IDEUtil.showErrorDialog(
						project,
						"Error adding plugin \"" + newPluginId + "\" to " + LivePluginAppComponent.pluginsRootPath(),
						addNewPluginTitle
				);
			}
			log.error(e);
		}

		RefreshPluginsPanelAction.refreshPluginTree();
	}

	public static class PluginIdValidator implements InputValidatorEx {
		private String errorText;

		@Override public boolean checkInput(String pluginId) {
			boolean isValid = !LivePluginAppComponent.pluginExists(pluginId);
			errorText = isValid ? null : "There is already a plugin with this name";
			return isValid;
		}

		@Nullable @Override public String getErrorText(String pluginId) {
			return errorText;
		}

		@Override public boolean canClose(String pluginId) {
			return true;
		}
	}
}

File: src/main/liveplugin/toolwindow/addplugin/AddNewPluginAction.kt
package liveplugin.toolwindow.addplugin

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.ui.InputValidatorEx
import com.intellij.openapi.ui.Messages
import liveplugin.Icons
import liveplugin.IdeUtil
import liveplugin.LivePluginAppComponent.Companion.pluginIdToPathMap
import liveplugin.LivePluginPaths
import liveplugin.LivePluginPaths.groovyExamplesPath
import liveplugin.LivePluginPaths.kotlinExamplesPath
import liveplugin.pluginrunner.groovy.GroovyPluginRunner
import liveplugin.pluginrunner.kotlin.KotlinPluginRunner
import liveplugin.toolwindow.RefreshPluginsPanelAction
import liveplugin.toolwindow.util.createFile
import liveplugin.toolwindow.util.readSampleScriptFile
import java.io.IOException

class AddNewGroovyPluginAction: AddNewPluginAction(
    text = "Groovy Plugin",
    description = "Create new Groovy plugin",
    scriptFileName = GroovyPluginRunner.mainScript,
    scriptFileText = readSampleScriptFile("$groovyExamplesPath/default-plugin.groovy")
)

class AddNewKotlinPluginAction: AddNewPluginAction(
    text = "Kotlin Plugin",
    description = "Create new Kotlin plugin",
    scriptFileName = KotlinPluginRunner.mainScript,
    scriptFileText = readSampleScriptFile("$kotlinExamplesPath/default-plugin.kts")
)

open class AddNewPluginAction(
    text: String,
    description: String,
    private val scriptFileName: String,
    private val scriptFileText: String
): AnAction(text, description, Icons.newPluginIcon), DumbAware {

    private val log = Logger.getInstance(AddNewPluginAction::class.java)
    private val addNewPluginTitle = "Add $text"

    override fun actionPerformed(event: AnActionEvent) {
        val project = event.project

        val newPluginId = Messages.showInputDialog(
            project,
            "Enter new plugin name:",
            addNewPluginTitle, null, "", PluginIdValidator()
        ) ?: return

        try {
            createFile("${LivePluginPaths.livePluginsPath}/$newPluginId", scriptFileName, scriptFileText, whenCreated = { virtualFile ->
                if (project != null) FileEditorManager.getInstance(project).openFile(virtualFile, true)
            })
        } catch (e: IOException) {
            if (project != null) IdeUtil.showErrorDialog(project, "Error adding plugin '$newPluginId' to ${LivePluginPaths.livePluginsPath}", addNewPluginTitle)
            log.error(e)
        }
        RefreshPluginsPanelAction.refreshPluginTree()
    }
}

class PluginIdValidator: InputValidatorEx {
    private var errorText: String? = null

    override fun checkInput(pluginId: String) =
        if (pluginId in pluginIdToPathMap().keys) {
            errorText = "There is already a plugin with this name"
            false
        } else {
            errorText = null
            true
        }

    override fun getErrorText(pluginId: String) = errorText

    override fun canClose(pluginId: String) = true
}
